### PR TITLE
fix: remove length parameters

### LINKED ISSUE #68: TMX Performance Issues due to Length Filter
Recently it seems like many people have complained about TMX's API performance causing RMX runs to hang without being able to load the next map. I would guess that the recently added preloading feature will help a lot in this regard, but I found another interesting detail which might help creating a fix or workaround.

If I'm not mistaken, the API call made during a TM2020 run is `https://trackmania.exchange/mapsearch2/search?api=on&random=1&etags=23,37,40&vehicles=1&mtype=TM_Race&lengthop=1&length=9`. Trying it out in my browser, the response either takes a very long time, or even times out completely. Removing the `lengthop` and `length` parameters or configuring them for a smaller range (e.g. 1 & 8) greatly reduces response latency for me. Curiously, even using `lengthop=3&length=8` which should functionally be the same (<9 vs <=8) improves latencies significantly for me.

I have no idea why this exact combination of parameters leads to slowdowns (maybe it also isn't special in any way and the cause is that it's the default parameter set used by the plugin and therefore many people), but playing around with the parameters might alleviate some of the issues (e.g. using the abovementioned 3 & 8). While I do not know how the TMX backend looks like, I would guess there is a chance that it is not purely a load issue but also some kind of configuration problem (e.g. missing/corrupted indices or stuff like that). From what I understood, you are acquainted with some of the people at TMX, so maybe this information can help the troubleshooting on their side.

### PR DESCRIPTION
**Pull request by EdVraz**
This allows you to disable the usage of length query parameters in API requests. This offers a partial fix for #68.

Ideally this also completely fixes the timeout issues, but I'm not sure if making query parameters opt-in immediately is the best thing to do.

### CODE REVIEW

File: src/Settings/SearchingParameters.as [hunk 1]
```diff
@@ -51,6 +70,10 @@ namespace PluginSettings
 
     bool initArrays = false;
 
+    // add a setting that streamers can toggle to switch back to the old length checks in case the TMX API starts failing again.
```
Comment (GreepTheSheep) [inline src/Settings/SearchingParameters.as:73]: anyone*

stop relying on streamers, it's useless

File: src/Settings/SearchingParameters.as [hunk 2]
```diff
@@ -69,10 +92,12 @@ namespace PluginSettings
 
     [SettingsTab name="Searching"]
     void RenderSearchingSettingTab()
-    {
-        CustomRules = UI::Checkbox("\\$fc0"+Icons::ExclamationTriangle+" \\$zAllow search parameters in RMC. Forbidden on official Leaderboard.", CustomRules);
-        UI::NewLine();
+    {   
+        UI::Text("Length filter setting. Toggle this when TMX gives super long response times.");
+        UseLengthChecksInRequests = UI::Checkbox("Use length filters in API requests", UseLengthChecksInRequests);
+        UI::Separator();
```
Comment (GreepTheSheep) [inline src/Settings/SearchingParameters.as:98]: Can you move this setting down ? Like moving to upper length selector (L117)

File: src/Utils/MX/Methods/LoadRandomMap.as [hunk 3]
```diff
@@ -35,6 +35,96 @@ namespace MX
...
+                    map.LengthName == "1 m 30 s" || map.LengthName == "1 m 45 s" || map.LengthName == "2 min" || map.LengthName == "2 m 30 s" || map.LengthName == "3 min"
+                )) {
+                    Log::Warn("Map is too long, retrying...");
+                    PreloadRandomMap();
+                    return;
+                }
+            } else if (PluginSettings::MapLength != "Anything") {
+                int requiredLength = PluginSettings::SearchingMapLengthsMilliseconds[PluginSettings::SearchingMapLengths.Find(PluginSettings::MapLength)];
+                switch (PluginSettings::SearchingMapLengthOperators.Find(PluginSettings::MapLengthOperator)) {
+                    case 0:  // exact is prorbably a bit too strict so we'll allow a 5 second difference
```
Comment (GreepTheSheep) [inline src/Utils/MX/Methods/LoadRandomMap.as:51]: 5 seconds is too short, can you try making 30 second difference ?

File: src/Utils/MX/Methods/LoadRandomMap.as [hunk 4]
```diff
@@ -35,6 +35,96 @@ namespace MX
...
+                        } else if (!(map.AuthorTime <= requiredLength)) {
+                            Log::Warn("Map is not the correct length, retrying...");
+                            PreloadRandomMap();
+                            return;
+                        }
+                        break;
+
+                    case 4:
+                        if (requiredLength == 100000000) {
+                            if (map.AuthorTime <= 300000) {
```
Comment (GreepTheSheep) [inline src/Utils/MX/Methods/LoadRandomMap.as:111]: As I said L53, 3min rule to a variable

**Review by GreepTheSheep — CHANGES_REQUESTED** [review]
Please check the comments

File: src/Utils/MX/Methods/LoadRandomMap.as (hunk 3 above)
Comment (EdVraz) [reply src/Utils/MX/Methods/LoadRandomMap.as:51]: 30 seconds seems a bit extreme considering the lowest configurable interval is 15 seconds, so for me it seems more logical to have it range within 15 seconds? With 30 seconds difference you can a 45 second map when you set only 15 second maps

Comment (GreepTheSheep) [reply src/Utils/MX/Methods/LoadRandomMap.as:51]: 15 seconds is ok 👍

File: src/Utils/MX/Methods/LoadRandomMap.as [hunk 5]
```diff
@@ -35,6 +35,96 @@ namespace MX
...
+                    Log::Warn("Map is too long, retrying...");
+                    PreloadRandomMap();
+                    return;
+                }
+            } else if (PluginSettings::MapLength != "Anything") {
+                int requiredLength = PluginSettings::SearchingMapLengthsMilliseconds[PluginSettings::SearchingMapLengths.Find(PluginSettings::MapLength)];
+                switch (PluginSettings::SearchingMapLengthOperators.Find(PluginSettings::MapLengthOperator)) {
+                    case 0:  // exact is prorbably a bit too strict so we'll allow a 5 second difference
+                        if (requiredLength == 100000000) {
+                            if (map.AuthorTime <= 300000+10000) {
```
[the remark this answers is not in the document]
Comment (GreepTheSheep) [reply src/Utils/MX/Methods/LoadRandomMap.as:53]: Ah I thought it was for 3 minutes, so that's ok

**Comment by EdVraz** [comment]
Alright, changed

**Review by GreepTheSheep — APPROVED** [review]

**Comment by Wain4** [comment]
"3 min" shouldn't be in the list of allowed lengths, the original map search returned times strictly less than 3 minutes.
(length 8 and 9 are "2 min 30 s" and "3 min", lengthop 1 and 3 are < and <=, the two versions of the query were 8/3 and 9/1)

**Pull request merged**